gpt.lua

```diff
@@ -0,0 +1,555 @@
+      if retries >= max_retries then
+        debug_log("Max retries reached.")
+        retries = 0
+        stop_polling(client)
+        return
+      else
+        retries = retries + 1
+      end
+    elseif current_status == Status.waiting then
+      update_progress_indicator()
+    end
+  end
+
+  if os.difftime(os.time(), start_time) >= timeout then
+    debug_log('Reached time limit of ' .. timeout .. ', stopping polling.')
+    retries = 0
+    stop_polling(client)
+    return
+  end
+end
+
+-- Sends json request to the remote service helper.
+local function send(request)
+  set_current_status(Status.waiting)
+  client = make_client()
+  start_time = os.time()
+  debug_log('Sending request... \n')
+  client:send(request)
+  poll(client)
+end
+
+-- Primary entrypoint to the script's functionality. Initiates a check
+-- of the UI to see if a supported written content item is being displayed.
+-- If so, then submit a request to the remote helper script.
+function fetch_generated_text()
+  skip = skip + 1
+  if skip < 20 then return end
+  skip = 0
+
+  if current_status ~= Status.start then
+    debug_log("Current status was not start status, aborting. Status was: " .. string_from_Status(current_status))
+    return
+  end
+
+  local knowledge_description, content_type = knowledge_description()
+
+  if knowledge_description == last_knowledge_description then
+    return
+  end
+
+  if not knowledge_description then
+    debug_log('Poem description became nil, retrying...')
+    last_knowledge_description = nil
+    return
+  end
+
+  if content_type == Content_Type.unsupported then
+    gui_text = "This content type is not supported. Please select a " .. valid_content_type_list .. "."
+    last_knowledge_description = nil
+    return
+  end
+
+  debug_log('Got new ' .. content_type .. " description: " .. knowledge_description .. "\n")
+  last_knowledge_description = knowledge_description
+  gui_text = "Generating text from description, please wait...  "
+  debug_log("Submitting request to OpenAI remote service... \n")
+  send(request_from(knowledge_description, content_type))
+end
+
+--
+-- GUI: Overlay
+--
+
+GPTBannerOverlay = defclass(GPTBannerOverlay, overlay.OverlayWidget)
+GPTBannerOverlay.ATTRS{
+    default_pos={x=-35,y=-2},
+    default_enabled=true,
+    viewscreens={'dwarfmode/ViewSheets/UNIT','dwarfmode/ViewSheets/ITEM'},
+    frame={w=30, h=1},
+    frame_background=gui.CLEAR_PEN,
+}
+
+function GPTBannerOverlay:init()
+    self:addviews{
+      widgets.TextButton{
+        frame={t=0, l=0},
+        label='AI Generation View',
+        key='CUSTOM_CTRL_G',
+        on_activate=function() view = view and view:raise() or GPTScreen{}:show() end,
+      },
+    }
+end
+
+function GPTBannerOverlay:onInput(keys)
+  if GPTBannerOverlay.super.onInput(self, keys) then return true end
+
+  if keys._MOUSE_R_DOWN or keys.LEAVESCREEN then
+    if view then
+      view:dismiss()
+    end
+  end
```

Comment:
do we actually need to close the window if the knowledge description screen is closed? why not just leave it open so the player can appreciate it? they can always close it with a right click on the window at any time.

Reply:
Thing is, right click usually doesn't close the knowledge description window, it just backs you to the list of known written content items, at which point we want the window closed so the user can open another knowledge item without necessarily generating it as soon as it's opened (which is what would happen if the window has remained opened). If someone is trying to economize which things they generate, that would be annoying. 

The good news is that if you click back into the same knowledge item you just left, the text is all still there, and you can always alt-tab to the python cmd console to see the output there also in case you forgot to screenshot a good one. 

I would suggest running the plugin and playing around with it, and then if you think I've misread the situation please let me know and I'll make the adjustment, or maybe we think of a third solution that's better than either of these options.